pkg/util/globalerror/grpc_test.go

```diff
@@ -284,3 +300,77 @@ func checkErrorWithStatusDetails(t *testing.T, details []any, expected *mimirpb.
 		require.Equal(t, expected, errDetails)
 	}
 }
+
+func TestGRPCClientClosingConnectionError(t *testing.T) {
+	ctx := context.Background()
+
+	_, client, cc := prepareTest(t)
+
+	// Calls to Succeed() should be successful when cc is open.
+	_, err := client.Succeed(ctx, nil)
+	require.NoError(t, err)
+
+	// We close cc.
+	err = cc.Close()
+	require.NoError(t, err)
+
+	// Calls to Succeed() should fail with "grpc: the client connection is closing" when cc is closed.
+	_, err = client.Succeed(ctx, nil)
+	checkGRPCConnectionIsClosingError(t, err)
+}
+
+func prepareTest(t *testing.T) (dskitserver.FakeServerServer, dskitserver.FakeServerClient, *grpc.ClientConn) {
+	grpcServer := grpc.NewServer()
+	t.Cleanup(grpcServer.GracefulStop)
+
+	server := &mockServer{}
+	dskitserver.RegisterFakeServerServer(grpcServer, server)
+
+	listener, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+
+	go func() {
+		require.NoError(t, grpcServer.Serve(listener))
+	}()
+
+	// Create a real gRPC client connecting to the gRPC server we control in this test.
+	clientCfg := grpcclient.Config{}
+	flagext.DefaultValues(&clientCfg)
+
+	opts, err := clientCfg.DialOption(nil, nil)
+	require.NoError(t, err)
+
+	cc, err := grpc.NewClient(listener.Addr().String(), opts...)
+	require.NoError(t, err)
+
+	client := dskitserver.NewFakeServerClient(cc)
+
+	// This is another source of "grpc: the client connection is closing",
+	// because at this point the connection is already closed.
+	t.Cleanup(func() {
+		err := cc.Close()
+		checkGRPCConnectionIsClosingError(t, err)
+	})
+
+	return server, client, cc
+}
+
+func checkGRPCConnectionIsClosingError(t *testing.T, err error) {
+	require.Error(t, err)
+	require.NotErrorIs(t, err, context.Canceled)
+
+	errWithCtx := WrapGRPCErrorWithContextError(err)
+	stat, ok := grpcstatus.FromError(errWithCtx)
+	require.True(t, ok)
+	require.Equal(t, codes.Canceled, stat.Code())
+	require.Equal(t, grpcClientConnectionIsClosingErr, stat.Message())
+	require.NotErrorIs(t, errWithCtx, context.Canceled)
+}
```

Comment:
I think the most important thing here is the last line:
```go
require.NotErrorIs(t, errWithCtx, context.Canceled)
```

But the method name doesn't reflect that. IMO, it makes harder to understand the test, because the test wasn't checking what "the error isn't context.Canceled".

OTOH, this test also does the wrapping itself, which is the other important part of the test.

In other words: I'd move the contents of this method to the test itself.

Reply:
Hi @colega, 
I have refactored the test. I kept the `checkGRPCConnectionIsClosing()` function, which actually checks that the passed error is a `grpc: the client connection is closing` error.
OTOH, I put the wrapping and the checks that both original and wrapped errors are not a `context cancled` error to the test itself.